pkg/tls/certificatemanagement/controller/ca.go

```diff
@@ -0,0 +1,533 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var (
+	log                     = logf.Log.WithName("tls")
+	errInvalidCertDNSNames  = errors.New("cert has the wrong DNS names")
+	errInvalidCertNoPEMData = errors.New("cert has no PEM data")
+)
+
+type trustedBundle struct {
+	certificateManager render.CertificateManager
+	pem                []byte
+	annotations        map[string]string
+}
+
+// CreateTrustedBundle creates a TrustedBundle, which provides standardized methods for mounting a bundle of certificates to trust.
+func CreateTrustedBundle(ca render.CertificateManager, certificates ...render.Certificate) render.TrustedBundle {
+	bundle := &trustedBundle{certificateManager: ca,
+		pem: []byte(fmt.Sprintf("# certificate name: %s\n%s\n\n", ca.Name(), string(ca.CertificatePEM()))),
+		annotations: map[string]string{
+			ca.HashAnnotationKey(): ca.HashAnnotationValue(),
+		}}
+	bundle.AddCertificates(certificates...)
+	return bundle
+}
+
+// AddCertificates Adds the PEM blocks of the certificates to the bundle.
+func (t *trustedBundle) AddCertificates(certificates ...render.Certificate) {
+	pemBuf := bytes.Buffer{}
+	pemBuf.Write(t.pem) // err is always nil.
+	for _, cert := range certificates {
+		if cert != nil && cert.CertificatePEM() != nil && !t.certificateManager.Issued(cert.CertificatePEM()) {
+			t.annotations[fmt.Sprintf("hash.operator.tigera.io/%s", cert.Name())] = rmeta.AnnotationHash(cert.CertificatePEM())
+			pemBuf.WriteString(fmt.Sprintf("# certificate name: %s\n%s\n\n",
+				cert.Name(), string(cert.CertificatePEM()))) // err is always nil
+		}
+	}
+	t.pem = pemBuf.Bytes()
+}
+
+func (t *trustedBundle) MountPath() string {
+	return render.TrustedCertBundleMountPath
+}
+
+func (t *trustedBundle) HashAnnotations() map[string]string {
+	return t.annotations
+}
+
+func (t *trustedBundle) VolumeMount() corev1.VolumeMount {
+	return corev1.VolumeMount{
+		Name:      render.TrustedCertConfigMapName,
+		MountPath: render.TrustedCertVolumeMountPath,
+		ReadOnly:  true,
+	}
+}
+
+func (t *trustedBundle) Volume() corev1.Volume {
+	return corev1.Volume{
+		Name: render.TrustedCertConfigMapName,
+		VolumeSource: corev1.VolumeSource{
+			ConfigMap: &corev1.ConfigMapVolumeSource{
+				LocalObjectReference: corev1.LocalObjectReference{Name: render.TrustedCertConfigMapName},
+			},
+		},
+	}
+}
+
+func (t *trustedBundle) ConfigMap(namespace string) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{Kind: "ConfigMap", APIVersion: "v1"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      render.TrustedCertConfigMapName,
+			Namespace: namespace,
+		},
+		Data: map[string]string{
+			render.TrustedCertConfigMapKeyName: string(t.pem),
+		},
+	}
+}
+
+// AddPEM Adds a PEM block to the bundle.
+func (t *trustedBundle) AddPEM(annotationName string, pem []byte) {
```

Comment:
You may have said something about this offline, but why is this needed as opposed to having a "CreateCertificateFromPem" which returns a certificate then adding this through the constructor?